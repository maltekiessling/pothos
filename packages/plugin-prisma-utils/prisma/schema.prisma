// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema
datasource db {
  provider = "postgresql"
  url      = "postgresql://prisma:prisma@localhost:5455/tests?schema=prisma-utils"
}

generator client {
  provider = "prisma-client-js"
  output   = "../tests/client"
}

generator pothos {
  provider     = "prisma-pothos-types"
  clientOutput = "../tests/client"
  output       = "../tests/generated.ts"
}

model Post {
  id         Int         @id @default(autoincrement())
  createdAt  DateTime    @unique @default(now())
  updatedAt  DateTime    @updatedAt
  title      String
  content    String?
  published  Boolean     @default(false)
  author     User        @relation(fields: [authorId], references: [id])
  comments   Comment[]
  authorId   Int
  media      PostMedia[]
  tags       String[]
  categories Category[]
}

enum Category {
  TECH
  SCIENCE
  CULTURE
  SPORTS
}

model Media {
  id           Int         @id @default(autoincrement())
  url          String
  posts        PostMedia[]
  uploadedBy   User?       @relation(fields: [uploadedById], references: [id])
  uploadedById Int?
}

model PostMedia {
  id      Int   @id @default(autoincrement())
  post    Post  @relation(fields: [postId], references: [id])
  media   Media @relation(fields: [mediaId], references: [id])
  postId  Int
  mediaId Int
  order   Int
}

model Comment {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @unique @default(now())
  content   String
  author    User     @relation(fields: [authorId], references: [id])
  post      Post     @relation(fields: [postId], references: [id])
  authorId  Int
  postId    Int
}

model Profile {
  id     Int     @id @default(autoincrement())
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model User {
  id        Int       @id @default(autoincrement())
  email     String    @unique
  name      String?
  posts     Post[]
  comments  Comment[]
  profile   Profile?
  followers Follow[]  @relation("followers")
  following Follow[]  @relation("following")
  Media     Media[]
}

model Follow {
  fromId Int
  toId   Int
  from   User @relation("following", fields: [fromId], references: [id])
  to     User @relation("followers", fields: [toId], references: [id])

  @@id([fromId, toId], name: "compositeID")
}

model Unrelated {
  id   Int     @id @default(autoincrement())
  name String?
}

model WithID {
  id                  String                @id
  FindUniqueRelations FindUniqueRelations[]
}

model WithUnique {
  id                  String                @unique
  FindUniqueRelations FindUniqueRelations[]
}

model WithCompositeID {
  a String
  b String

  FindUniqueRelations FindUniqueRelations[]

  @@id([a, b])
}

model WithCompositeUnique {
  a String
  c String? @unique
  b String

  FindUniqueRelations FindUniqueRelations[]

  @@unique([a, c])
  @@unique([a, b])
}

model FindUniqueRelations {
  id String @id

  withID_id String
  withID    WithID @relation(fields: [withID_id], references: [id])

  withUnique_id String
  withUnique    WithUnique @relation(fields: [withUnique_id], references: [id])

  withCompositeID_a String
  withCompositeID_b String
  withCompositeID   WithCompositeID @relation(fields: [withCompositeID_a, withCompositeID_b], references: [a, b])

  withCompositeUnique_a String
  withCompositeUnique_b String
  withCompositeUnique   WithCompositeUnique @relation(fields: [withCompositeUnique_a, withCompositeUnique_b], references: [a, b])
}
