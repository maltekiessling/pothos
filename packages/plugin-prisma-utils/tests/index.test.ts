import { execute, printSchema } from 'graphql';
import { gql } from 'graphql-tag';
import { prisma } from './examples/simple/builder';
import schema from './examples/simple/schema';

let queries: unknown[] = [];
prisma.$use((params, next) => {
  queries.push(params);

  return next(params);
});

describe('prisma utils', () => {
  afterEach(() => {
    queries = [];
  });

  it('generates schema', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('returns filtered posts', async () => {
    const query = gql`
      query {
        posts(
          order: { author: { name: Asc, profile: null } }
          filter: { id: { not: { equals: 23752, not: null } } }
        ) {
          id
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
    });

    expect(result).toMatchInlineSnapshot(`
      {
        "data": {
          "posts": [
            {
              "id": "26",
            },
            {
              "id": "27",
            },
            {
              "id": "28",
            },
          ],
        },
      }
    `);

    expect(queries).toMatchInlineSnapshot(`
      [
        {
          "action": "findMany",
          "args": {
            "orderBy": {
              "author": {
                "name": "asc",
                "profile": undefined,
              },
            },
            "take": 3,
            "where": {
              "id": {
                "not": {
                  "equals": 23752,
                  "not": undefined,
                },
              },
            },
          },
          "dataPath": [],
          "model": "Post",
          "runInTransaction": false,
        },
      ]
    `);
  });
});
