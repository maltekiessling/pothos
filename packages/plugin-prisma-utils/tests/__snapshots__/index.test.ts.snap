// Vitest Snapshot v1

exports[`prisma utils > generates schema 1`] = `
"input CommentFilter {
  createdAt: DateTime
}

input CommentListFilter {
  some: CommentFilter
}

scalar DateTime

input IntFilter {
  equals: Int
  not: IntFilter
}

type Mutation {
  createUser(data: UserCreate!): User!
}

enum OrderBy {
  Asc
  Desc
}

type Post {
  id: ID!
}

input PostCreate {
  title: String!
}

input PostFilter {
  author: UserFilter

  \\"\\"\\"filter by author id\\"\\"\\"
  authorId: IntFilter
  comments: CommentListFilter
  createdAt: DateTime
  id: IntFilter
  title: String
}

input PostOrderBy {
  author: UserOrderBy
  createdAt: OrderBy
  id: OrderBy
  title: OrderBy
}

input PostWhereUnique {
  id: Int
}

input ProfileOrderBy {
  bio: OrderBy
  user: UserOrderBy
}

type Query {
  post(list: StringListFilter, title: StringFilter): Post
  posts(filter: PostFilter, order: PostOrderBy): [Post!]!
}

input StringFilter {
  contains: String
  equals: String
  not: StringFilter
  startsWith: String
}

input StringListFilter {
  every: StringFilter
  none: StringFilter
  some: StringFilter
}

type User {
  email: String!
  id: ID!
  posts: [Post!]!
}

input UserCreate {
  email: String!
  name: String
  posts: UserCreatePostsRelation
}

input UserCreatePostsRelation {
  connect: [PostWhereUnique!]
  create: [PostCreate!]
}

input UserFilter {
  NOT: UserFilter
  id: IntFilter
}

input UserOrderBy {
  name: OrderBy
  profile: ProfileOrderBy
}"
`;
